src/Polysemy/MTL.hs

```diff
@@ -0,0 +1,189 @@
+import           Data.Kind (Type, Constraint)
+
+import           Polysemy
+import           Polysemy.Reader
+import           Polysemy.Writer
+import           Polysemy.State
+import           Polysemy.Error
+
+{- | Open type-family mapping a single constraint of the form @(Type -> Type) -> Constraint@, e.g., @MonadState s@,
+to a polysemy effect which can be used to re-interpret that constraint, e.g., 'State s'.
+-}
+type family CanonicalEffect (p :: (Type -> Type) -> Constraint) :: (Type -> Type) -> Type -> Type
+
+type instance  CanonicalEffect (S.MonadReader env) = Reader env
+type instance  CanonicalEffect (S.MonadWriter w)   = Writer w
+type instance  CanonicalEffect (S.MonadState s)    = State s
+type instance  CanonicalEffect (S.MonadError e)    = Error e
+
+
+{- | A newtype wrapper for a monadic action, parameterized by
+a constraint, @p@ on a @(Type -> Type)@ (e.g., a monad); @m@, a specific
+@(Type -> Type)@; and a polysemy effect type-list @r@. With "Data.Reflection"
+we can create instances of @p (ConstrainedAction p m r)@ using functions from
+@Sem r@.
+-}
+newtype ConstrainedAction (p :: (Type -> Type) -> Constraint) (m :: Type -> Type) (s :: Type) (x :: Type)
+  = ConstrainedAction
+    { action :: m x
+    } deriving (Functor, Applicative, Monad)
+
+{- | For a constraint to be "absorbable" by @Sem r@, there needs to be an instance of this class,
+containing the dictionary signatures as a record of functions and the
+reflected entailment of @p (ConstrainedAction p m r)@ from the reified dictionary.
+-}
+class ReifiableConstraint1 p where
+  data Dict (p :: (Type -> Type) -> Constraint) (m :: Type -> Type)
+  reifiedInstance :: Monad m => R.Reifies s (Dict p m) :- p (ConstrainedAction p m s)
+
+{- | This class contains an instance of the dictionary for some set of effects
+parameterized by a polysemt effect list @r@.
+Typically, you would write this instance for any @r@
+satisfying the constraint that the "canonical" effect is a member.  But you
+could also use it to discharge constraints which require multiple polysemy effects.
+-}
+class ReifiableConstraint1 p => SemDictionary p r where
+  semDictionary :: Dict p (Sem r)
+
+-- | Given a reifiable constraint, and a dictionary to use, discharge the constraint.
+using :: forall p m a. (Monad m, ReifiableConstraint1 p) => Dict p m -> (p m => m a) -> m a
+using d m =
+  R.reify d $ \(_ :: Proxy s) -> m \\ C.trans (C.unsafeCoerceConstraint :: (p (ConstrainedAction p m s) :- p m)) reifiedInstance
+
+{- | Given a "canonical" dictionary for @p@ using the polysemy effects in @r@,
+discharge the constraint @p@.
+-}
+absorb :: forall p r a. SemDictionary p r => (p (Sem r) => Sem r a) -> Sem r a
+absorb = (using @p) semDictionary
+
+------------------------------------------------------------------------------
+absorbReader :: forall i r a. Member (Reader i) r => (S.MonadReader i (Sem r) => Sem r a) -> Sem r a
+absorbReader f = absorb @(S.MonadReader i) f --using (MonadReader ask local) f
```

Comment:
The polysemy-plugin combined with the fundep on mtl should make specifying `i` here (and in general, all ty params on the mtl effects) unnecessary.  Presumably this would typecheck as:

```haskell
absorbReader = absorb @(S.MonadReader _)
```

or maybe even as

```haskell
absorbReader = absorb
```

Reply:
If this is the case, you can drop the `forall` on the typesig.